lib/zip_source_zip.c

```diff
@@ -49,8 +51,14 @@ ZIP_EXTERN zip_source_t *zip_source_zip_create(zip_t *srcza, zip_uint64_t srcidx
         flags |= ZIP_FL_COMPRESSED;
     else
         flags &= ~ZIP_FL_COMPRESSED;
-    
-    return _zip_source_zip_new(srcza, srcidx, flags, start, (zip_uint64_t)len, NULL, error);
+
+    password = srcza->default_password;
+
+    if (password != NULL && password[0] == '\0') {
+        password = NULL;
+    }
+
+    return _zip_source_zip_new(srcza, srcidx, flags, start, (zip_uint64_t)len, password, error);
```

Comment:
`zip_set_default_password` already turns an empty password into `NULL`, so the check here is redundant. Just pass `srcza->default_password` to `_zip_source_zip_new`.